### PR TITLE
Python: Relax restriction of flow through `async with`

### PR DESCRIPTION
**Pull request by RasmusWL**
/cc @pwntester who pointed out this problem

### CODE REVIEW

File: python/ql/lib/semmle/python/dataflow/new/internal/DataFlowPrivate.qll [hunk 1]
```diff
@@ -304,8 +304,12 @@ module EssaFlow {
       // see `with_flow` in `python/ql/src/semmle/python/dataflow/Implementation.qll`
       with.getContextExpr() = contextManager.getNode() and
       with.getOptionalVars() = var.getNode() and
-      not with.isAsync() and
       contextManager.strictlyDominates(var)
+      // note: we allow this for both `with` and `async with`, since some
+      // implementations do `async def __aenter__(self): return self`, so you can do
+      // both:
+      // * `foo = x.foo(); await foo.async_methoid(); foo.close()` and
```
Comment (tausbn) [inline python/ql/lib/semmle/python/dataflow/new/internal/DataFlowPrivate.qll:311]:
```suggestion
      // * `foo = x.foo(); await foo.async_method(); foo.close()` and
```

File: python/ql/lib/semmle/python/dataflow/new/internal/DataFlowPrivate.qll [hunk 2]
```diff
@@ -314,6 +318,12 @@ module EssaFlow {
     //  nodeTo is `x`, essa var
     //
     // This makes the cfg node the local source of the awaited value.
+    //
+    // We have this step in addition to the step above, to handle cases where the QL
+    // modeling of `f(42)` requires a `.getAwaited()` step (in API graphs) when not
+    // using `async with`, so you can do both:
+    // * `foo = await x.foo(); await foo.async_methoid(); foo.close()` and
```
Comment (tausbn) [inline python/ql/lib/semmle/python/dataflow/new/internal/DataFlowPrivate.qll:325]:
```suggestion
    // * `foo = await x.foo(); await foo.async_method(); foo.close()` and
```

**Review by tausbn — DISMISSED** [review]
Two minor typo fixes, otherwise this looks good to me. 👍

**Review by tausbn — APPROVED** [review]

**Comment by RasmusWL** [comment]
Performance results didn't show anything interesting.

**Pull request merged**